File: Algorithms_python/8_String_to_Integer(atoi).py
class Solution:
    def myAtoi(self, s: str) -> int:
        i = 0
        n = len(s)
        if n == 0:
            return 0
        while s[i] == ' ':
            i += 1
            if i == n:
                return 0
        if s[i] == '-':
            sign = -1
        else:
            sign = 1
        if s[i] in ['-', '+']:
            i += 1
        res = 0
        flag = (2 ** 31 - 1) // 10
        while i < n:
            if not s[i].isdigit():
                break
            c = int(s[i])
            if res > flag or (res == flag and c > 7):
                return 2 ** 31 - 1 if sign > 0 else -2 ** 31
            res = res * 10 + c
            i += 1
        return sign * res
'''
输入的顺序是固定的，必定是空格，符号，数字，再是迷惑的字母
所以按输入的顺序暴力检索，先查空格，再查符号存着，然后查数字
记得不能超过范围，空字符串，只有空格的字符串等等限制，对着test测就完了
用动态规划写，其实思路是一模一样的，时间空间复杂度也都是O(m,n).
'''

File: Algorithms_C++/8_String_to_Integer(atoi).cpp
class Solution {
public:
    int myAtoi(string s) {
        unsigned long len = s.length();
        int index = 0;
        while (index < len){
            if (s[index] != ' '){
                break;
            }
            ++index;
        }
        if (index == len){
            return 0;
        }
        int sign = 1;
        if (s[index] == '+'){
            ++index;
        }
        else if(s[index] == '-'){
            sign = -1;
            ++index;
        }

        int res = 0;
        while (index < len){
            char curChar = s[index];
            if (curChar < '0' || curChar > '9'){
                break;
            }
            if (res > INT_MAX / 10 || (res == INT_MAX / 10 && (curChar - '0') > INT_MAX % 10)) {
                return INT_MAX;
            }
            if (res < INT_MIN / 10 || (res == INT_MIN / 10 && (curChar - '0') > -(INT_MIN % 10))) {
                return INT_MIN;
            }
            res = res * 10 + sign * (curChar - '0');
            index++;
        }
        return res;
    }
};